### DISCUSSION TITLE
received fatal signal 11 (Segmentation fault)

### DISCUSSION THREAD

**Discussion by artur5236** [addiejanssen/domoticz-solaredge-modbustcp-plugin#28] [Q&A]
Hello,
I have below error:
"Error: Domoticz(pid:3698, tid:3727('SE6K-modbus')) received fatal signal 11 (Segmentation fault)"

details:
2022-02-02 12:58:59.071 Status: SE6K-modbus: (SE6K-modbus) Started.
2022-02-02 12:58:59.071 Status: SE6K-modbus: (SE6K-modbus) Entering work loop.
2022-02-02 12:58:59.637 Status: SE6K-modbus: (SE6K-modbus) Initialized version 1.0.6, author 'Addie Janssen'
2022-02-02 12:59:29.047 SE6K-modbus: (SE6K-modbus) Pushing 'onHeartbeatCallback' on to queue
2022-02-02 12:59:29.097 (SE6K-modbus) Processing 'onHeartbeatCallback' message
2022-02-02 12:59:29.097 SE6K-modbus: (SE6K-modbus) Calling message handler 'onHeartbeat'.
2022-02-02 12:59:29.097 SE6K-modbus: (SE6K-modbus) onHeartbeat
2022-02-02 12:59:29.097 SE6K-modbus: (SE6K-modbus) Retrying to communicate with inverter after: 2022-02-02 13:00:59.734652
2022-02-02 12:59:59.025 SE6K-modbus: (SE6K-modbus) Pushing 'onHeartbeatCallback' on to queue
2022-02-02 12:59:59.075 (SE6K-modbus) Processing 'onHeartbeatCallback' message
2022-02-02 12:59:59.075 SE6K-modbus: (SE6K-modbus) Calling message handler 'onHeartbeat'.
2022-02-02 12:59:59.075 SE6K-modbus: (SE6K-modbus) onHeartbeat
2022-02-02 12:59:59.076 SE6K-modbus: (SE6K-modbus) Retrying to communicate with inverter after: 2022-02-02 13:00:59.734652
2022-02-02 13:07:28.290 SE6K-modbus hardware started.
2022-02-02 13:07:28.551 SE6K-modbus: (SE6K-modbus) Debug logging mask set to: PYTHON PLUGIN QUEUE IMAGE DEVICE CONNECTION MESSAGE ALL
2022-02-02 13:07:28.551 SE6K-modbus: (SE6K-modbus) onStart Address: 192.168.1.80 Port: 1502
2022-02-02 13:07:28.856 SE6K-modbus: (SE6K-modbus) Connection established with: 192.168.1.80:1502
2022-02-02 13:07:28.856 SE6K-modbus: (SE6K-modbus) Inverter type: sunspecDID.THREE_PHASE_INVERTER
2022-02-02 13:07:28.290 Status: SE6K-modbus: (SE6K-modbus) Entering work loop.
2022-02-02 13:07:28.290 Status: SE6K-modbus: (SE6K-modbus) Started.
2022-02-02 13:07:28.550 Status: SE6K-modbus: (SE6K-modbus) Initialized version 1.0.6, author 'Addie Janssen'
2022-02-02 13:07:28.856 Error: Domoticz(pid:3698, tid:3727('SE6K-modbus')) received fatal signal 11 (Segmentation fault)
2022-02-02 13:13:59.232 SE6K-modbus hardware started.
2022-02-02 13:13:59.478 SE6K-modbus: (SE6K-modbus) Debug logging mask set to: PYTHON PLUGIN QUEUE IMAGE DEVICE CONNECTION MESSAGE ALL
2022-02-02 13:13:59.478 SE6K-modbus: (SE6K-modbus) onStart Address: 192.168.1.80 Port: 1502
2022-02-02 13:13:59.786 SE6K-modbus: (SE6K-modbus) Connection established with: 192.168.1.80:1502
2022-02-02 13:13:59.786 SE6K-modbus: (SE6K-modbus) Inverter type: sunspecDID.THREE_PHASE_INVERTER
2022-02-02 13:13:59.232 Status: SE6K-modbus: (SE6K-modbus) Entering work loop.
2022-02-02 13:13:59.232 Status: SE6K-modbus: (SE6K-modbus) Started.
2022-02-02 13:13:59.477 Status: SE6K-modbus: (SE6K-modbus) Initialized version 1.0.6, author 'Addie Janssen'
2022-02-02 13:13:59.786 Error: Domoticz(pid:1749, tid:1777('SE6K-modbus')) received fatal signal 11 (Segmentation fault)

**Comment by artur5236** [comment]
Which version of solaredge-modbus this plugin works fine?

**Reply by addiejanssen** [reply]
solaredge_modbus 0.6.2 works for sure.
I tested with later versions and have not found issues so far.

**Comment by artur5236** [comment]
I tested, solaredge_modbus 0.6.2, 0.6.6 and 0.70 and neither version worked. 
I have Domoticz version 2021.1 and rpi os with Debian version: 11 (bullseye)

**Reply by addiejanssen** [reply]
I wonder if this a plugin issue. I see many similar issues with different plugins and all sorts of software conflicts.
So, the plugin may trigger the error but is not the actual cause of the error itself.
Now, debugging this is going to be very hard.

Do you have other Python based plugins running in your Domoticz setup?
This post might be interesting: https://www.domoticz.com/forum/viewtopic.php?t=35340
They found that there was a specific issue cause by the Python version they used.

I also see posts where multiple Python based plugins influence each other; one way to test if that is the case, is by disabling all other plugins for a moment and see if the error remains.

**Comment by addiejanssen** [comment]
Do you have a file named domoticz_crash.log file in the domoticz folder?
If so, can you share it here?

**Comment by artur5236** [comment]
domoticz_crash.log

`
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/arm-linux-gnueabihf/libthread_db.so.1".
0xb6a314d4 in __GI___wait4 (pid=3615, stat_loc=0xbed92ee0, options=0, usage=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:27
27      ../sysdeps/unix/sysv/linux/wait4.c: No such file or directory.
  Id   Target Id                                  Frame
* 1    Thread 0xb5f30040 (LWP 3427) "domoticz"    0xb6a314d4 in __GI___wait4 (pid=3615, stat_loc=0xbed92ee0, options=0, usage=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:27
  2    Thread 0xb5a23200 (LWP 3428) "Watchdog"    0xb6a2af8c in __GI___clock_nanosleep_time64 (clock_id=clock_id@entry=0, flags=flags@entry=0, req=0xb5a22b70, req@entry=0xb5a22b68, rem=0xb5a22b80, rem@entry=0xb5a22b78) at ../sysdeps/unix/sysv/linux/clock_nanosleep.c:52
  3    Thread 0xb5222200 (LWP 3429) "SQLHelper"   syscall () at ../sysdeps/unix/sysv/linux/arm/syscall.S:37
  4    Thread 0xb00ff200 (LWP 3439) "MainWorker"  __pthread_clockjoin_ex (threadid=2846847488, thread_return=0x0, clockid=clockid@entry=0, abstime=abstime@entry=0x0, block=block@entry=true) at pthread_join_common.c:145
  5    Thread 0xa9af7200 (LWP 3472) "SE6K-modbus" __GI___clock_gettime64 (clock_id=clock_id@entry=5, tp=0xa9af6b10, tp@entry=0xa9af6b08) at ../sysdeps/unix/sysv/linux/clock_gettime.c:52

Thread 5 (Thread 0xa9af7200 (LWP 3472) "SE6K-modbus"):
#0  __GI___clock_gettime64 (clock_id=clock_id@entry=5, tp=0xa9af6b10, tp@entry=0xa9af6b08) at ../sysdeps/unix/sysv/linux/clock_gettime.c:52
#1  0xb6a2aad0 in __GI___clock_gettime (clock_id=clock_id@entry=5, tp=0xa9af6b34, tp@entry=0xa9af6b2c) at ../sysdeps/unix/sysv/linux/clock_gettime.c:82
#2  0xb6a1ecc8 in time (timer=0x0) at ../time/time.c:26
#3  0x00649b6c in Plugins::CPlugin::Do_Work() ()
#4  0x009a15a4 in execute_native_thread_routine ()
#5  0xb6ba6300 in start_thread (arg=0xa9af7200) at pthread_create.c:477
#6  0xb6a6a208 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73 from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 4 (Thread 0xb00ff200 (LWP 3439) "MainWorker"):
#0  __pthread_clockjoin_ex (threadid=2846847488, thread_return=0x0, clockid=clockid@entry=0, abstime=abstime@entry=0x0, block=block@entry=true) at pthread_join_common.c:145
#1  0xb6ba779c in __pthread_join (threadid=<optimized out>, thread_return=<optimized out>) at pthread_join.c:24
#2  0x009a17e8 in std::thread::join() ()
#3  0x006472ac in Plugins::CPlugin::StopHardware() ()
#4  0x0041dd34 in CDomoticzHardwareBase::Stop() ()
#5  0x0023351c in MainWorker::RemoveDomoticzHardware(CDomoticzHardwareBase*) ()
#6  0x002565cc in MainWorker::AddHardwareFromParams(int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, _eHardwareTypes, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned short, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int, int, int, int, int, int, bool) ()
#7  0x00259384 in MainWorker::RestartHardware(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ()
#8  0x00259b98 in MainWorker::Do_Work() ()
#9  0x009a15a4 in execute_native_thread_routine ()
#10 0xb6ba6300 in start_thread (arg=0xb00ff200) at pthread_create.c:477
#11 0xb6a6a208 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73 from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 3 (Thread 0xb5222200 (LWP 3429) "SQLHelper"):
#0  syscall () at ../sysdeps/unix/sysv/linux/arm/syscall.S:37
#1  0x0097f014 in std::__atomic_futex_unsigned_base::_M_futex_wait_until(unsigned int*, unsigned int, bool, std::chrono::duration<long long, std::ratio<1ll, 1ll> >, std::chrono::duration<long long, std::ratio<1ll, 1000000000ll> >) ()
#2  0x002a31ac in CSQLHelper::Do_Work() ()
#3  0x009a15a4 in execute_native_thread_routine ()
#4  0xb6ba6300 in start_thread (arg=0xb5222200) at pthread_create.c:477
#5  0xb6a6a208 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73 from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 2 (Thread 0xb5a23200 (LWP 3428) "Watchdog"):
#0  0xb6a2af8c in __GI___clock_nanosleep_time64 (clock_id=clock_id@entry=0, flags=flags@entry=0, req=0xb5a22b70, req@entry=0xb5a22b68, rem=0xb5a22b80, rem@entry=0xb5a22b78) at ../sysdeps/unix/sysv/linux/clock_nanosleep.c:52
#1  0xb6a2b080 in __GI___clock_nanosleep (clock_id=clock_id@entry=0, flags=flags@entry=0, req=<optimized out>, rem=0xb5a22bb0) at ../sysdeps/unix/sysv/linux/clock_nanosleep.c:92
#2  0xb6a31830 in __GI___nanosleep (requested_time=<optimized out>, remaining=<optimized out>) at nanosleep.c:27
#3  0x0021f744 in sleep_milliseconds(long) ()
#4  0x0028b880 in Do_Watchdog_Work() ()
#5  0x009a15a4 in execute_native_thread_routine ()
#6  0xb6ba6300 in start_thread (arg=0xb5a23200) at pthread_create.c:477
#7  0xb6a6a208 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73 from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 1 (Thread 0xb5f30040 (LWP 3427) "domoticz"):
#0  0xb6a314d4 in __GI___wait4 (pid=3615, stat_loc=0xbed92ee0, options=0, usage=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:27
#1  0x0028b14c in dumpstack_gdb(bool) ()
#2  0x0028b718 in signal_handler(int, siginfo_t*, void*) ()
#3  <signal handler called>
#4  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#5  0xb69a5804 in __GI_abort () at abort.c:79
#6  0xb5c1b254 in ?? () from /lib/arm-linux-gnueabihf/libpython3.9.so
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Main thread:
#0  0xb6a314d4 in __GI___wait4 (pid=3615, stat_loc=0xbed92ee0, options=0, usage=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:27
#1  0x0028b14c in dumpstack_gdb(bool) ()
#2  0x0028b718 in signal_handler(int, siginfo_t*, void*) ()
#3  <signal handler called>
#4  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#5  0xb69a5804 in __GI_abort () at abort.c:79
#6  0xb5c1b254 in ?? () from /lib/arm-linux-gnueabihf/libpython3.9.so
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
[Inferior 1 (process 3427) detached]

`

**Reply by addiejanssen** [reply]
Looking at this, I don't think it is an issue with the plugin itself nor with the modbus library.
Which version of Python are you on?

**Comment by artur5236** [comment]
I used Python 3.9.2

Currently, modbus library (solaredge-modbus 0.7.0) works fine:
`
python3 example.py 192.168.1.80 1502
Inverter(192.168.1.80:1502, connectionType.TCP: timeout=1, retries=3, unit=0x1):

Registers:
        Manufacturer: SolarEdge
        Model: SE6K-RW0TEBNN4
        Type: Three Phase Inverter
        Version: 0004.0014.0107
        Serial: ######
        Status: Producing
        Temperature: 19.48°C
        Current: 2.20A
        Phase 1 Current: 0.72A
        Phase 2 Current: 0.74A
        Phase 3 Current: 0.73A
        Phase 1 voltage: 409.40V
        Phase 2 voltage: 409.40V
        Phase 3 voltage: 406.80V
        Phase 1-N voltage: 235.60V
        Phase 2-N voltage: 237.50V
        Phase 3-N voltage: 234.50V
        Frequency: 50.02Hz
        Power: 477.50W
        Power (Apparent): 520.90VA
        Power (Reactive): -208.15VAr
        Power Factor: -91.66%
        Total Energy: 72105Wh
        DC Current: 0.65A
        DC Voltage: 750.20V
        DC Power: 484.70W
`

**Reply by addiejanssen** [reply]
That's good news. I am pretty sure it is not the modbus library, and I think it is also not related to my plugin.

I have been looking at the domoticz issues here:
https://github.com/domoticz/domoticz/issues?q=is%3Aissue+is%3Aopen+%22signal+11%22
And for similar issues that have been closed.
Now, there have been code changes related to some of the Python and segmentation fault issues, but they will not be in the 2021.1 version.
Is it an option for you to upgrade to Domoticz 2022.1? That got released a few days ago.
Not sure this will fix the issue, but it should include at least the patches that fixed some of the closed issues related to Python and some segmentation faults.

**Comment by artur5236** [comment]
I have upgraded Domoticz to latest version and after enabling plugin, domoticz crashed.

**Reply by dthery** [reply]
i have the same issu and I can't find a solution :-/

**Reply by dthery** [reply]
I just solved by installing the beta version of Domoticz.
